### PR TITLE
Update `kubectl wait` command to handle multiple conditions correctly

### PR DESCRIPTION
**Pull request by seshachalam-yv**
**How to categorize this PR?**

/area testing
/kind bug test

**What this PR does / why we need it**:

- This PR updates the `kubectl wait` command used to check conditions for a seed to handle multiple conditions correctly. 
- **In the previous version, each `--for` option would override the previous one.**  So always checks last conditions only.
```
kubectl wait --for=condition=gardenletready --for=condition=extensionsready --for=condition=bootstrapped seed local2 --timeout=5m
```
- This PR resolves this issue by using separate `kubectl wait` commands for each condition, rather than relying on multiple `--for` options.
- This ensures that all conditions are properly checked and eliminates any risk of the conditions being overwritten or ignored.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
cc: @rfranzke 
**Release note**:

```other operator
NONE
```

### CODE REVIEW

**Comment by rfranzke** [comment]
I don't understand the rationale of this - can you explain again? Doesn't this now drastically prolong the execution time if something does not work as expected? Generally, `5m` should be enough for the seed to come up. When is this not sufficient?

**Comment by rfranzke** [comment]
Nice finding, I guess we weren't aware of this. There is also https://github.com/kubernetes/kubernetes/issues/95759 👀

**Comment by rfranzke** [comment]
Instead of waiting for each conditions separately for at most `5-10m`, can we follow this approach https://github.com/gardener/gardener/blob/9ff1e217c662817f44ec4e4c5c4042b42ff2dd4a/skaffold.yaml#L1123-L1130

For example, by doing this

```bash
#!/usr/bin/env bash

for i in `seq 1 60`; do
  seed_name="local"

  if [ "$(kubectl get seed "$seed_name" -o jsonpath='{.status.conditions[?(@.type=="GardenletReady")].status}')" == "True" ] && \
     [ "$(kubectl get seed "$seed_name" -o jsonpath='{.status.conditions[?(@.type=="Bootstrapped")].status}')" == "True" ] && \
     [ "$(kubectl get seed "$seed_name" -o jsonpath='{.status.conditions[?(@.type=="ExtensionsReady")].status}')" == "True" ]
  then
    echo "Seed has become ready!"
    break
  fi

  echo "Wait until seed gets ready"
  sleep 5
done
```

File: hack/ci-e2e-kind-upgrade.sh [hunk 1]
```diff
@@ -162,9 +162,21 @@ function set_seed_name() {
...
-  echo "Wait until seed gets upgraded from version '$GARDENER_PREVIOUS_RELEASE' to '$GARDENER_NEXT_RELEASE'"
-  kubectl wait seed $1 --timeout=5m \
-    --for=jsonpath='{.status.gardener.version}'=$GARDENER_NEXT_RELEASE && condition=gardenletready && condition=extensionsready && condition=bootstrapped
+  echo "Wait until seed '$1' gets upgraded from version '$GARDENER_PREVIOUS_RELEASE' to '$GARDENER_NEXT_RELEASE'"
+  for i in $(seq 1 60); do
+    if [ "$(kubectl get seed "$1" -o jsonpath='{.status.conditions[?(@.type=="GardenletReady")].status}')" == "True" ] && \
+      [ "$(kubectl get seed "$1" -o jsonpath='{.status.conditions[?(@.type=="Bootstrapped")].status}')" == "True" ] && \
+      [ "$(kubectl get seed "$1" -o jsonpath='{.status.conditions[?(@.type=="SeedSystemComponentsHealthy")].status}')" == "True" ] && \
+      [ "$(kubectl get seed "$1" -o jsonpath='{.status.conditions[?(@.type=="ExtensionsReady")].status}')" == "True" ] && \
+      [ "$(kubectl get seed "$1" -o jsonpath='{.status.conditions[?(@.type=="BackupBucketsReady")].status}')" == "True" ]
```
Comment (oliver-goetz) [inline hack/ci-e2e-kind-upgrade.sh]: Nit:
```suggestion
  for _ in $(seq 1 60); do
    if [ "$(kubectl get seed "$1" -o jsonpath='{.status.conditions[?(@.type=="GardenletReady")].status}')" == "True" ] && \
       [ "$(kubectl get seed "$1" -o jsonpath='{.status.conditions[?(@.type=="Bootstrapped")].status}')" == "True" ] && \
       [ "$(kubectl get seed "$1" -o jsonpath='{.status.conditions[?(@.type=="SeedSystemComponentsHealthy")].status}')" == "True" ] && \
       [ "$(kubectl get seed "$1" -o jsonpath='{.status.conditions[?(@.type=="ExtensionsReady")].status}')" == "True" ] && \
       [ "$(kubectl get seed "$1" -o jsonpath='{.status.conditions[?(@.type=="BackupBucketsReady")].status}')" == "True" ]
```
Same for all other occurrences.

**Comment by oliver-goetz** [comment]
There is one more thing. 
`kubectl wait` returns an error when the condition timed out. At the moment the script continues after 5 minutes no matter if the conditions are true or not.

**Comment by seshachalam-yv** [comment]
> There is one more thing. `kubectl wait` returns an error when the condition timed out. At the moment the script continues after 5 minutes no matter if the conditions are true or not.

@oliver-goetz  Script will immediately exit if some error because we used `bash -ec` and `set -o errexit`

File: skaffold.yaml [hunk 2]
```diff
@@ -1167,15 +1172,19 @@ profiles:
...
       echo "Wait until seed is ready"
-      for i in `seq 1 30`;
-      do
-        if kubectl --kubeconfig=$GARDENER_LOCAL_KUBECONFIG get seed local2 2> /dev/null; then
+      for _ in $(seq 1 60); do
+        seed_name="local2"
+
+        if [ "$(kubectl get seed "$seed_name" -o jsonpath='{.status.conditions[?(@.type=="GardenletReady")].status}')" == "True" ] &&
+          [ "$(kubectl get seed "$seed_name" -o jsonpath='{.status.conditions[?(@.type=="Bootstrapped")].status}')" == "True" ] &&
+          [ "$(kubectl get seed "$seed_name" -o jsonpath='{.status.conditions[?(@.type=="ExtensionsReady")].status}')" == "True" ]; then
```
Comment (rfranzke) [inline skaffold.yaml]: Here (https://github.com/gardener/gardener/pull/7439/files#diff-cb88f836d7d2eec7bccbed73c47a452b8eb3e376b02bae73ecae449bba96214aR1126-R1130, you are checking more conditions. Why not here as well?

**Comment by oliver-goetz** [comment]
> > There is one more thing. `kubectl wait` returns an error when the condition timed out. At the moment the script continues after 5 minutes no matter if the conditions are true or not.
> 
> @oliver-goetz Script will immediately exit if some error because we used `bash -ec` and `set -o errexit`

There is nothing which raises an error when the loop was running 60 times and the conditions are still false. Thus, the script would continue, wouldn't it?

File: skaffold.yaml [hunk 3]
```diff
@@ -1198,15 +1207,16 @@ profiles:
     - -ec
     - |
       echo "Wait until seed is ready"
-      for i in `seq 1 60`;
-      do
-        if kubectl --kubeconfig=$GARDENER_LOCAL_KUBECONFIG get seed $SEED_NAME 2> /dev/null; then
+      for _ in $(seq 1 60); do
+        if [ "$(kubectl get seed "$SEED_NAME" -o jsonpath='{.status.conditions[?(@.type=="GardenletReady")].status}')" == "True" ] &&
+          [ "$(kubectl get seed "$SEED_NAME" -o jsonpath='{.status.conditions[?(@.type=="Bootstrapped")].status}')" == "True" ]; then
```
Comment (rfranzke) [inline skaffold.yaml]: Same here.

File: docs/deployment/getting_started_locally.md [hunk 4]
```diff
@@ -97,7 +97,9 @@ Afterwards, the Gardener resources will be deployed into the cluster.
 You can wait for the `Seed` to be ready by running:
 
 ```bash
-kubectl wait --for=condition=gardenletready --for=condition=extensionsready --for=condition=bootstrapped seed local --timeout=5m
+kubectl wait --for=condition=gardenletready seed local --timeout=5m
+kubectl wait --for=condition=extensionsready seed local --timeout=5m
+kubectl wait --for=condition=bootstrapped seed local --timeout=5m
```
Comment (rfranzke) [inline docs/deployment/getting_started_locally.md]: What about all these - shouldn't they be adapted as well just like in `skaffold.yaml`? 

Also, there are similar statements in https://github.com/gardener/gardener/blob/master/docs/development/getting_started_locally.md (maybe more, please check).

Comment (rfranzke) [inline docs/deployment/getting_started_locally.md]: What about all these - shouldn't they be adapted as well just like in `skaffold.yaml`? 

Also, there are similar statements in https://github.com/gardener/gardener/blob/master/docs/development/getting_started_locally.md (maybe more, please check).

**Review by rfranzke — CHANGES_REQUESTED** [review]

**Comment by seshachalam-yv** [comment]
> > > There is one more thing. `kubectl wait` returns an error when the condition timed out. At the moment the script continues after 5 minutes no matter if the conditions are true or not.
> > 
> > 
> > @oliver-goetz Script will immediately exit if some error because we used `bash -ec` and `set -o errexit`
> 
> There is nothing which raises an error when the loop was running 60 times and the conditions are still false. Thus, the script would continue, wouldn't it?

Yes, you are right, still loop continues if some kubectl error like `The connection to the server localhost:8080 was refused - did you specify the right host or port?`

I think, still okay to continue.
WDYT

File: docs/deployment/getting_started_locally.md (hunk 4 above)
Comment (rfranzke) [reply docs/deployment/getting_started_locally.md]: This is not yet addressed.

Comment (seshachalam-yv) [reply docs/deployment/getting_started_locally.md]: Thanks @rfranzke fixed it.

**Comment by oliver-goetz** [comment]
> > > > There is one more thing. `kubectl wait` returns an error when the condition timed out. At the moment the script continues after 5 minutes no matter if the conditions are true or not.
> > > 
> > > 
> > > @oliver-goetz Script will immediately exit if some error because we used `bash -ec` and `set -o errexit`
> > 
> > 
> > There is nothing which raises an error when the loop was running 60 times and the conditions are still false. Thus, the script would continue, wouldn't it?
> 
> Yes, you are right, still loop continues if some kubectl error like `The connection to the server localhost:8080 was refused - did you specify the right host or port?`
> 
> I think, still okay to continue. WDYT

We should cover the case when the seed is not available after 60 loop runs yet and exit with an error in this case.
That's easier to check than tests which might time out because the seed is not ready.

File: docs/deployment/getting_started_locally.md (hunk 4 above)
Comment (rfranzke) [reply docs/deployment/getting_started_locally.md]: No, it's still not addressed. We should not have these many `kubectl wait` statements - we should use a similar loop like in the `skaffold.yaml`. Probably this should be put into a script in the `hack` folder and then users can call it from there to check if a shoot or seed is ready?

File: hack/usage/wait.sh [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+while [ "${retries}" -lt "${RETRY_LIMIT}" ]; do
+  if [ -z "$NAMESPACE" ]; then
+    CONDITION_STATES=$(kubectl get "${RESOURCE_TYPE}" "${OBJECT_NAME}" -o jsonpath='{.status.conditions}')
+  else
+    CONDITION_STATES=$(kubectl get "${RESOURCE_TYPE}" "${OBJECT_NAME}" -n "$NAMESPACE" -o jsonpath='{.status.conditions}')
+  fi
+  ALL_PASSED=true
+
+  for condition in "${CONDITIONS[@]}"; do
+    if ! echo "${CONDITION_STATES}" | jq -e '.[] | select(.type == "'"${condition}"'").status == "True"' > /dev/null; then
```
Comment (oliver-goetz) [inline hack/usage/wait.sh]: The kind jobs are failing because `jq` is not installed. You could replace `jq` with `yq` which is already installed in our tool chain.

File: example/gardener-local/kind/cluster/templates/cluster.yaml [hunk 6]
```diff
@@ -1,5 +1,7 @@
 apiVersion: kind.x-k8s.io/v1alpha4
 kind: Cluster
+featureGates:
+  ExpandedDNSConfig: true
```
Comment (oliver-goetz) [inline example/gardener-local/kind/cluster/templates/cluster.yaml:7]: Why is this needed? I'm just wondering because it is not mentioned anywhere.

**Review by oliver-goetz — COMMENTED** [review]
It is a good idea to create a script for this 😄 
Please replace `jq` by `yq` that it works in the kind container too.

File: hack/usage/wait.sh [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+GREEN='\033[0;32m'
+NO_COLOR='\033[0m'
+
+echo "Checking conditions for ${RESOURCE_TYPE}/${OBJECT_NAME}..."
+retries=0
+while [ "${retries}" -lt "${RETRY_LIMIT}" ]; do
+  if [ -z "$NAMESPACE" ]; then
+    CONDITION_STATES=$(kubectl get "${RESOURCE_TYPE}" "${OBJECT_NAME}" -o jsonpath='{.status.conditions}')
+  else
+    CONDITION_STATES=$(kubectl get "${RESOURCE_TYPE}" "${OBJECT_NAME}" -n "$NAMESPACE" -o jsonpath='{.status.conditions}')
```
Comment (oliver-goetz) [inline hack/usage/wait.sh]: Afaik `yq` works on yaml only, so this should be adapted too.

File: example/gardener-local/kind/cluster/templates/cluster.yaml (hunk 6 above)
Comment (seshachalam-yv) [reply example/gardener-local/kind/cluster/templates/cluster.yaml:7]: Multi-zone upgrade tests are failing due to this https://github.com/gardener/gardener/issues/7297, our e2e kind cluster   version is `v1.24.7`, so we must enable featureGate `ExpandedDNSConfig`

File: skaffold.yaml [hunk 8]
```diff
@@ -1165,17 +1154,7 @@ profiles:
...
-      for i in `seq 1 30`;
-      do
-        if kubectl --kubeconfig=$GARDENER_LOCAL_KUBECONFIG get seed local2 2> /dev/null; then
-          break
-        fi
-        echo "Wait until seed gets created by gardenlet"
-        sleep 2
-      done
-      kubectl --kubeconfig=$GARDENER_LOCAL_KUBECONFIG wait --for=condition=gardenletready --for=condition=extensionsready --for=condition=bootstrapped seed local2 --timeout=5m
+    - hack/usage/wait.sh seed local2 GardenletReady Bootstrapped SeedSystemComponentsHealthy ExtensionsReady BackupBucketsReady
```
Comment (oliver-goetz) [inline skaffold.yaml]: The `hack/usage/wait.sh` script probably needs a `KUBECONFIG` parameter. The script needs to query the `garden cluster` but in this case the kind2 cluster, for which the deployment is running, is not the garden cluster, but a standalone seed.
This makes the migration tests failing.

File: skaffold.yaml [hunk 9]
```diff
@@ -1196,17 +1175,7 @@ profiles:
...
-      for i in `seq 1 60`;
-      do
-        if kubectl --kubeconfig=$GARDENER_LOCAL_KUBECONFIG get seed $SEED_NAME 2> /dev/null; then
-          break
-        fi
-        echo "Wait until seed gets created by gardenlet"
-        sleep 2
-      done
-      kubectl --kubeconfig=$GARDENER_LOCAL_KUBECONFIG wait --for=condition=gardenletready --for=condition=bootstrapped seed $SEED_NAME --timeout=5m
+    - hack/usage/wait.sh seed "$SEED_NAME" GardenletReady Bootstrapped SeedSystemComponentsHealthy ExtensionsReady BackupBucketsReady
```
Comment (oliver-goetz) [inline skaffold.yaml]: Same here regarding the `KUBECONFIG` parameter. Additionally, this scenario does not deploy any BackupBucket by default.

**Review by oliver-goetz — COMMENTED** [review]
I might have found the reason why the migration tests are currently failing.

File: skaffold.yaml (hunk 8 above)
Comment (oliver-goetz) [reply skaffold.yaml]: This might work too
``` suggestion
    - KUBECONFIG=$GARDENER_LOCAL_KUBECONFIG hack/usage/wait.sh seed local2 GardenletReady Bootstrapped SeedSystemComponentsHealthy ExtensionsReady BackupBucketsReady
```
You could use this everywhere in skaffold.yaml to make the kubeconfig explicit.

File: skaffold.yaml (hunk 9 above)
Comment (oliver-goetz) [reply skaffold.yaml]:
```suggestion
    - KUBECONFIG=$GARDENER_LOCAL_KUBECONFIG hack/usage/wait.sh seed "$SEED_NAME" GardenletReady Bootstrapped SeedSystemComponentsHealthy ExtensionsReady
```

**Comment by oliver-goetz** [comment]
/lgtm

@rfranzke please add your final review.

File: docs/deployment/getting_started_locally.md [hunk 10]
```diff
@@ -94,12 +94,6 @@ Afterwards, the Gardener resources will be deployed into the cluster.
 
 ## Creating a `Shoot` Cluster
 
-You can wait for the `Seed` to be ready by running:
-
-```bash
-kubectl wait --for=condition=gardenletready --for=condition=extensionsready --for=condition=bootstrapped seed local --timeout=5m
-```
-
```
Comment (rfranzke) [inline docs/deployment/getting_started_locally.md]: Shouldn't we add an example for the `wait.sh` script to wait for the seed to be ready?
If not then adapt the next sentence (this doesn't read well anymore without this).

Actually, you have done this in the dev/ document.

File: docs/deployment/getting_started_locally.md [hunk 11]
```diff
@@ -195,12 +189,6 @@ make gardenlet-kind2-up
 
 The following steps assume that you are using the kubeconfig that points to the `gardener-local` cluster (first KinD cluster): `export KUBECONFIG=example/gardener-local/kind/local/kubeconfig`.
 
-You can wait for the `local2` `Seed` to be ready by running:
-
-```bash
-kubectl wait --for=condition=gardenletready --for=condition=extensionsready --for=condition=bootstrapped seed local2 --timeout=5m
-```
-
```
Comment (rfranzke) [inline docs/deployment/getting_started_locally.md]: same

File: example/gardener-local/kind/cluster/templates/cluster.yaml (hunk 6 above)
Comment (rfranzke) [reply example/gardener-local/kind/cluster/templates/cluster.yaml:7]: Can you add a comment to it so that we don't forget about it?

File: hack/ci-e2e-kind-upgrade.sh [hunk 12]
```diff
@@ -194,7 +188,9 @@ make test-pre-upgrade GARDENER_PREVIOUS_RELEASE=$GARDENER_PREVIOUS_RELEASE GARDE
 
 echo "Upgrading gardener version '$GARDENER_PREVIOUS_RELEASE' to '$GARDENER_NEXT_RELEASE'"
 upgrade_to_next_release
-wait_until_seed_gets_upgraded "$SEED_NAME"
+
+echo "Wait until seed '$SEED_NAME' gets upgraded from version '$GARDENER_PREVIOUS_RELEASE' to '$GARDENER_NEXT_RELEASE'"
+./hack/usage/wait.sh seed "$SEED_NAME" GardenletReady Bootstrapped SeedSystemComponentsHealthy ExtensionsReady BackupBucketsReady
```
Comment (rfranzke) [inline hack/ci-e2e-kind-upgrade.sh]: Apart from waiting for the conditions, do we also wait for the `.status.gardener.version` to be updated?

File: hack/usage/wait.sh [hunk 13]
```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env bash
```
Comment (rfranzke) [inline hack/usage/wait.sh]: Nit: Can you call this script `wait-for.sh`?

File: hack/usage/wait.sh [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script waits until all conditions are passed for a given resource in a kubernetes cluster.
+# It takes the resource type, object name, and a list of conditions as arguments
+set -euo pipefail
+
+if [ "$#" -lt 3 ]; then
+  echo "Usage: $0 <resource_type> <object_name> <condition_1> <condition_2> ... <condition_n>
+Note: Namespace/RETRY_LIMIT will be used from the 'NAMESPACE'/'RETRY_LIMIT' environment variable if set, otherwise it is optional.
```
Comment (rfranzke) [inline hack/usage/wait.sh]:
```suggestion
Note: Namespace/RETRY_LIMIT will be used from the 'NAMESPACE'/'RETRY_LIMIT' environment variables if set, otherwise it is optional.
```

**Comment by rfranzke** [comment]
Also, I'm wondering why the upgrade tests still failed. Is there even more to it?

File: example/gardener-local/kind/cluster/templates/cluster.yaml [hunk 15]
```diff
@@ -1,5 +1,10 @@
 apiVersion: kind.x-k8s.io/v1alpha4
 kind: Cluster
+featureGates:
+# Introudced due to https://github.com/gardener/gardener/issues/7297#issuecomment-1377515385.
```
Comment (rfranzke) [inline example/gardener-local/kind/cluster/templates/cluster.yaml]: typo

File: example/gardener-local/kind/cluster/templates/cluster.yaml [hunk 16]
```diff
@@ -1,5 +1,10 @@
...
+# Feature gate 'ExpandedDNSConfig' will be true by default from k8s version 1.26.
+# Remove this once kind cluster upgraded to k8s version 1.26
```
Comment (rfranzke) [inline example/gardener-local/kind/cluster/templates/cluster.yaml]:
```suggestion
# Remove this once kind cluster version is upgraded to 1.26
```

File: hack/ci-e2e-kind-upgrade.sh [hunk 17]
```diff
@@ -194,7 +188,13 @@ make test-pre-upgrade GARDENER_PREVIOUS_RELEASE=$GARDENER_PREVIOUS_RELEASE GARDE
 
 echo "Upgrading gardener version '$GARDENER_PREVIOUS_RELEASE' to '$GARDENER_NEXT_RELEASE'"
 upgrade_to_next_release
-wait_until_seed_gets_upgraded "$SEED_NAME"
+
+echo "Wait until seed '$SEED_NAME' gets upgraded from version '$GARDENER_PREVIOUS_RELEASE' to '$GARDENER_NEXT_RELEASE'"
+kubectl wait seed $SEED_NAME --timeout=5m --for=jsonpath="{.status.gardener.version}=$GARDENER_NEXT_RELEASE"
+# RETRY_LIMIT has been increased to 240 (20 minutes) due to the upgrading of Gardener for seed.
+# In a single-zone setup, 3 istio-ingressgateway pods will be running, and it will take 9 minutes to complete the rollout.
+# In a multi-zone setup, 6 istio-ingressgateway pods will be running, and it will take 18 minutes to complete the rollout.
```
Comment (rfranzke) [inline hack/ci-e2e-kind-upgrade.sh]: Where are these values coming from? 😕 Why does it take `9m`/`18m`? Can we speed this up?

Also, the upgrade tests still fail, why?

Comment (seshachalam-yv) [reply hack/ci-e2e-kind-upgrade.sh]: > Where are these values coming from? 😕 Why does it take 9m/18m? Can we speed this up?

While upgrading Gardener, each `istio-ingressgateway` pod was healthy after 3 mins, so rollout to complete we need `9m`/ `18m` for `single/muti-zone` setups.
@ScheererJ  Is it possible to reduce time?

> Also, the upgrade tests still fail, why?
There is a downtime for while upgrading Gardener, I'm investigating more

Comment (ScheererJ) [reply hack/ci-e2e-kind-upgrade.sh]: We configure `istio` per default with update strategy:

```
  strategy:
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 0
```

Hence, the pods are individually rolled.

A single-zone setup has two pods per default. Each of them should take less than 3 minutes to be restarted. On a CI system it might be different due to higher general load, but 9 minutes seem like a high value, nevertheless.

In the multi-zone setup, we have four `istio` gateway namespaces. One has 6 replicas, the others have two per default. All four namespaces will be rolled in parallel, but the one with 6 replicas will likely determine the overall wall clock time. 18 minutes feels like a high water mark regarding the update.

It would probably be possible to speed this up if we change the update strategy in the local setup. The `PodDisruptionBudget` allows for a faster update.

File: hack/ci-e2e-kind-upgrade.sh [hunk 18]
```diff
@@ -194,9 +188,15 @@ make test-pre-upgrade GARDENER_PREVIOUS_RELEASE=$GARDENER_PREVIOUS_RELEASE GARDE
 
 echo "Upgrading gardener version '$GARDENER_PREVIOUS_RELEASE' to '$GARDENER_NEXT_RELEASE'"
 upgrade_to_next_release
-wait_until_seed_gets_upgraded "$SEED_NAME"
+
+echo "Wait until seed '$SEED_NAME' gets upgraded from version '$GARDENER_PREVIOUS_RELEASE' to '$GARDENER_NEXT_RELEASE'"
+kubectl wait seed $SEED_NAME --timeout=5m --for=jsonpath="{.status.gardener.version}=$GARDENER_NEXT_RELEASE"
+# RETRY_LIMIT has been increased to 240 (20 minutes) due to the upgrading of Gardener for seed.
+# In a single-zone setup, 3 istio-ingressgateway pods will be running, and it will take 9 minutes to complete the rollout.
```
Comment (ScheererJ) [inline hack/ci-e2e-kind-upgrade.sh]:
```suggestion
# In a single-zone setup, 2 istio-ingressgateway pods will be running, and it will take 9 minutes to complete the rollout.
```

File: hack/ci-e2e-kind-upgrade.sh [hunk 19]
```diff
@@ -183,7 +177,7 @@ kind_up
 trap "
 ( rm -rf $GARDENER_RELEASE_DOWNLOAD_PATH/gardener-releases);
 ( export_logs '$CLUSTER_NAME'; export_events_for_kind '$CLUSTER_NAME'; export_events_for_shoots )
-( kind_down;)
```
Comment (ScheererJ) [inline hack/ci-e2e-kind-upgrade.sh]: Is this left in debugging? I understand that you want to keep things running to debug what went wrong, but do we really want to change this default?

File: hack/ci-e2e-kind-upgrade.sh [hunk 20]
```diff
@@ -194,9 +188,15 @@ make test-pre-upgrade GARDENER_PREVIOUS_RELEASE=$GARDENER_PREVIOUS_RELEASE GARDE
...
+# RETRY_LIMIT has been increased to 240 (20 minutes) due to the upgrading of Gardener for seed.
+# In a single-zone setup, 3 istio-ingressgateway pods will be running, and it will take 9 minutes to complete the rollout.
+# In a multi-zone setup, 6 istio-ingressgateway pods will be running, and it will take 18 minutes to complete the rollout.
+RETRY_LIMIT=240 ./hack/usage/wait-for.sh seed "$SEED_NAME" GardenletReady Bootstrapped SeedSystemComponentsHealthy ExtensionsReady BackupBucketsReady
 
 echo "Running gardener post-upgrade tests"
 make test-post-upgrade GARDENER_PREVIOUS_RELEASE=$GARDENER_PREVIOUS_RELEASE GARDENER_NEXT_RELEASE=$GARDENER_NEXT_RELEASE
 
-gardener_down
+# gardener_down
```
Comment (ScheererJ) [inline hack/ci-e2e-kind-upgrade.sh]: Same as above: Is this intentional or just debugging code left in the PR?

File: hack/ci-e2e-kind-upgrade.sh [hunk 21]
```diff
@@ -194,9 +188,15 @@ make test-pre-upgrade GARDENER_PREVIOUS_RELEASE=$GARDENER_PREVIOUS_RELEASE GARDE
 
 echo "Upgrading gardener version '$GARDENER_PREVIOUS_RELEASE' to '$GARDENER_NEXT_RELEASE'"
 upgrade_to_next_release
-wait_until_seed_gets_upgraded "$SEED_NAME"
+
+echo "Wait until seed '$SEED_NAME' gets upgraded from version '$GARDENER_PREVIOUS_RELEASE' to '$GARDENER_NEXT_RELEASE'"
+kubectl wait seed $SEED_NAME --timeout=5m --for=jsonpath="{.status.gardener.version}=$GARDENER_NEXT_RELEASE"
+# RETRY_LIMIT has been increased to 240 (20 minutes) due to the upgrading of Gardener for seed.
```
Comment (ScheererJ) [inline hack/ci-e2e-kind-upgrade.sh]: It would be better if the internal retry logic would not be exposed to the caller here. The caller wants to set a timeout, but has to calculate with the internal 5 second sleep delay to get the retry limit.
What do you think about changing it to a more relatable value like seconds or minutes?

**Review by ScheererJ — CHANGES_REQUESTED** [review]

File: hack/ci-e2e-kind-upgrade.sh (hunk 20 above)
Comment (seshachalam-yv) [reply hack/ci-e2e-kind-upgrade.sh]: Yes, I added for debugging purpose.
Now reverted it.

File: hack/ci-e2e-kind-upgrade.sh (hunk 19 above)
Comment (seshachalam-yv) [reply hack/ci-e2e-kind-upgrade.sh]: Yes, I added for debugging purpose.
Now reverted it.

File: hack/ci-e2e-kind-upgrade.sh (hunk 21 above)
Comment (seshachalam-yv) [reply hack/ci-e2e-kind-upgrade.sh]: Thanks @ScheererJ for the suggestion.
I have fixed it so that the caller can now set the TIMEOUT in seconds. Operation will retry until the timeout is reached, with a 5-second sleep interval between each retry.

**Review by ScheererJ — APPROVED** [review]
/lgtm
